src/measure_theory/measure/finite_measure_weak_convergence.lean

```diff
@@ -0,0 +1,281 @@
+
+TODO:
+* Define the topologies on the above types.
+
+## Main results
+
+None yet.
+
+TODO:
+* Portmanteau theorem.
+
+## Notations
+
+No new notation is introduced.
+
+## Implementation notes
+
+The topology of weak convergence of finite Borel measures will be defined using a mapping from
+`finite_measures α` to `weak_dual ℝ≥0 (α →ᵇ ℝ≥0)`, inheriting the topology from there.
+
+## References
+
+* [Billingsley, *Convergence of probability measures*][billingsley1999]
+
+## Tags
+
+weak convergence of measures, finite measure, probability measure
+
+-/
+
+noncomputable theory
+open measure_theory
+open set
+open filter
+open_locale topological_space ennreal nnreal
+
+section finite_measures
+/-!
+### Finite measures
+In this section, we define the `Type` of finite measures on a given measurable space. This type
+will be equipped with the topology of weak convergence of measures when the measurable space is
+a topological space equipped with its Borel sigma-algebra.
+-/
+
+variables {α : Type} [measurable_space α]
+
+/-- Finite measures are defined as the subtype of measures that have the property of being finite
+measures (i.e., their total mass is finite). -/
+def finite_measures (α : Type*) [measurable_space α] : Type :=
+{ μ : measure α // finite_measure μ }
+
+namespace finite_measures
+
+instance has_zero {α : Type*} [measurable_space α] :
+  has_zero (finite_measures α) :=
+⟨{ val := 0,
+   property := measure_theory.finite_measure_zero, }⟩
+
+instance : inhabited (finite_measures α) := { default := 0 }
+
+instance {α : Type*} [measurable_space α] : has_add (finite_measures α) :=
+{ add := (λ (μ ν : finite_measures α),
+  { val := μ.val + ν.val,
+    property := @measure_theory.finite_measure_add α _ μ.val ν.val μ.prop ν.prop, }), }
+
+instance {α : Type*} [measurable_space α] : has_scalar ℝ≥0 (finite_measures α) :=
+{ smul := (λ (c : ℝ≥0) (μ : finite_measures α),
+  { val := c • μ.val,
+    property := @measure_theory.finite_measure_smul_nnreal α _ μ.val μ.prop c, }), }
+
+-- TODO: A coercion `finite_measures α → measure α` is crucial. Now it is the subtype coercion.
+-- Another alternative is to use a bijection with `vector_measure α ℝ≥0` as an intermediate step.
+-- Potential advantages of the vector measure alternative:
+--  * Structures of additive commutative monoid and ℝ≥0 module on `finite_measure α` could be
+--    "inherited".
+-- Potential drawbacks of the vector measure alternative:
+--  * No integration theory directly.
+--  * Monotonicity lost as non-measurable sets defined to have measure 0.
+
+/-- A finite measure can be interpreted as a measure. -/
+instance : has_coe (finite_measures α) (measure_theory.measure α) := coe_subtype
+
+instance to_measure.finite_measure (μ : finite_measures α) :
+  finite_measure (μ : measure α) := μ.prop
+
+instance (α : Type*) [measurable_space α] :
+  has_coe_to_fun (finite_measures α) := ⟨(λ _, set α → ℝ≥0),
+    (λ μ, (λ s, (μ.val.measure_of s).to_nnreal))⟩
```

Comment:
```suggestion
instance (α : Type*) [measurable_space α] :
  has_coe_to_fun (finite_measures α) :=
⟨λ _, set α → ℝ≥0, λ μ s, (μ s).to_nnreal⟩
```
This is probably nicer

Reply:
Here my preference would be the original, since the hover and "Go to definition" give more information about it, and the lenght difference is minor. Of course if there is an established style recommendation, I will follow it.

I implemented your other suggestions, they were clear improvements! Thank you very much!

Reply:
I'm thinking that there might be more lemmas for `(μ : measure α) s` than `μ.val.measure_of s`, in which case using  `(μ : measure α) s`  might be better since we can rewrite using these lemmas

Reply:
I see! I don't have a lot of experience about how much difference it makes, but this sounds like an issue.

(In fact the `to_nnreal` probably interferes a bit with existing lemmas anyway. I suppose this is an argument in favor of going via `finite_measure`, which has API. But having to provide proofs of measurability to use monotonicity would be a downside of that.)

I think I'm now in favor of your suggestion here, but I will wait a bit for any further comments before making the change.

Thank you again!

Reply:
Yes, I agree with Jason's comment here.  If there are several ways something could be phrased (like `μ.val` vs `(μ : measure α)`), it helps to write all the theory in terms of one fixed canonical phrasing, and we generally choose the coercion to be that canonical phrasing.